### ISSUE TITLE
ycecream: a no dependency. Pythonic fork of IceCream

### ISSUE THREAD

**Issue by salabim** [gruns/icecream#58]
I like the IceCream package very much, but there were several reasons why I made a fork.
This solves also a number of issues raised here.

The ycecream package can be found here: https://github.com/salabim/ycecream and can be installed directly from PyPI.

Here are the main differences:
* ycecream can't colourize the output (a nice feature of IceCream)
* ycecream runs only on Python 3.6 and higher. (IceCream runs even on Python 2.7).
* ycecream uses y as the standard interface, whereas IceCream uses ic. For compatibility, ycecream also supports ic.
* yceceam has no dependencies. IceCream on the other hand has many (asttoken, colorize, pyglets, ...).
* ycecream is just one .py file, whereas IceCream consists of a number of .py files. That makes it possible to use ycecream without even (pip) installing it. Just copy ycecream.py to your work directory.
* ycecream can be used as a decorator of a function or method showing the enter and/or exit event as well as the duration (useful for simple benchmarks)
* ycecream can output directly to a callable (like IceCream), but also an open file (e.g. sys.stdout) or a file given by the name (strr of Path) 
* ycecream uses a different PEP8 compatible API to customize (rather than IceCream's configureOutput method)
* ycecream time showing can be controlled independently from context showing
* ycecream can optionally show a delta (time since start of the program)
* ycecream does not sort dicts by default. This behaviour can be controlled with the sort_dict parameter. (This is implemented by including the pprint 3.8 source code)
* ycecream uses pytest for the test scripts rather than IceCream's unittest script.

**Comment by alexmojaki** [comment]
> I like the IceCream package very much, but there were several reasons why I made a fork.
> This solves also a number of issues raised here.

If an issue was raised here, why not make a PR solving that issue? If there was no issue here, why not make one? Seems like all of these features could be in icecream:

> * ycecream can be used as a decorator of a function or method showing the enter and/or exit event as well as the duration (useful for simple benchmarks)
> * ycecream can output directly to a callable (like IceCream), but also an open file (e.g. sys.stdout) or a file given by the name (strr of Path)
> * ycecream time showing can be controlled independently from context showing
> * ycecream can optionally show a delta (time since start of the program)
> * ycecream uses pytest for the test scripts rather than IceCream's unittest script.

As for this:

> * yceceam has no dependencies. IceCream on the other hand has many (asttoken, colorize, pyglets, ...).
> * ycecream is just one .py file, whereas IceCream consists of a number of .py files. That makes it possible to use ycecream without even (pip) installing it. Just copy ycecream.py to your work directory.

Perhaps this should be done in an automated manner with a tool like https://pypi.org/project/stickytape/ . Is that what https://github.com/salabim/ycecream/blob/main/install%20ycecream%20from%20github.py is? What will you do when there are changes to executing, asttokens, or icecream?

**Comment by salabim** [comment]
I started with this project as I didn't like the API (not Pythonic) and wanted a single source package. That didn't seemed to be just a PR for IceCream, so I started my own fork.

Feel free to use whatever you want from my package to make IceCream more useful (particularly decorators!).

With respect to changes in the underlying packages: these will of course not automatically be incorporated. On the other hand, ycecream will not suffer from changes in the API's of these packages (although that's quite unlikely).

The file 'install ycecream from github.py' is just another way
to install ycecream.py directly in site-packages, wthout PyPI.

**Comment by gruns** [comment]
> ycecream can be used as a decorator of a function or method showing the enter and/or exit event as well as the duration (useful for simple benchmarks)

this is definitely something we want. see https://github.com/gruns/icecream/issues/32

design discussion, and later pull requests, welcome there!

> ycecream uses a different PEP8 compatible API to customize (rather than IceCream's configureOutput method)

ya. as icecream has grown, so too have the number of options

it's now rather clunky to pack so many params into `ic.configureOutput()` and configuration in general could use a simplification and overhaul

> ycecream can output directly to a callable (like IceCream), but also an open file (e.g. sys.stdout) or a file given by the name (strr of Path)

this is certainly possible with `ic.configureOutput(outputFunction=yourFunctionHere)`

it could be useful to have those 'batteries' included and if you hand outputFunction a file object or string it 'does the right thing'. would be better to rename `outputFunction` to `output` in this case and `output` could take:

  1. a function: current implementation
  2. a string: open the file with the given string filepath and write to it
  3. file object: write to the file object using the file object API
 
---

@alexmojaki https://pypi.org/project/stickytape/ is neat. i, personally, havent seen it used by other projects, though

think it'd be worth using stickytape to build a single `icecream.py` file for download and inclusion?

at this time, my gut says no. this is the first time someone has requested a single file version of icecream

**Comment by alexmojaki** [comment]
I only found stickytape just now after a quick google. Seems like there's other possibilities for single file packaging.

You can see places it's used [here](https://github.com/mwilliamson/stickytape/network/dependents), the most significant being https://github.com/doyensec/inql with 704 stars where it looks like it works very well.

Overall it's not clear to me why this kind of thing is important. In the case of InQL it *looks* like you need a single file to load as an extension for other software. For icecream, I don't know what significant advantage there is. I've seen this kind of thing requested once before [here](https://github.com/cool-RR/PySnooper/issues/18), and again it's not clear if it was necessary.

I think the most sensible path would be to:

1. Integrate as much code as possible (within reason) into the base repo. icecream is popular and well known and contributions here will reach the most people and do the most good. Making ycecream equally well known will be very hard. Even if it happened, many will still use icecream anyway.
2. Keep the separate fork for differences that are not easily integrated, like breaking compatibility changes. But if it's very similar to the base repo because of the previous point, updates made in the base will be easy to merge into the fork.
3. Automate the process of building a single file with stickytape or something else. Maybe have it in the base repo, maybe not. Then updates in dependencies can still be used automatically. New Python versions tend to break asttokens and executing, so *some* strategy for keeping these up to date is necessary.

[I've done something similar before](https://github.com/alexmojaki/snoop/wiki/Comparison-to-PySnooper#why-not-add-these-features-to-pysnooper). I know having your own codebase and complete control is easier and more fun. But I still contributed as much as I could to the original repo before I had to make a fork, and it was worth it. And in general, code in my fork was not pulled back into the base by anyone. That doesn't happen easily or naturally.

**Comment by jkyeung** [comment]
@alexmojaki - Reading your comments makes me think perhaps you have glossed over (if not completely missed) an important point: ycecream doesn't have any dependencies.

This is not merely an easier or lower-tech way to achieve single-file packaging. It is a feature unto itself.

I was excited to try icecream until I discovered that it has quite many dependencies and indeed has a relatively large overall footprint, most of which is devoted to something I am actually not very interested in (colorizing).

I actually came on here to post a question or feature request issue (can we make the colorizing optional, as well as all the dependencies that are there just to support it?) and instead found that someone else not only had similar ideas, but actually built something! ycecream is almost exactly what I am looking for[1].

I don't think the motivation for ycecream has anything whatsoever to do with having complete control over a codebase. It's that salabim specifically wanted a simpler, smaller project with a different API. Those things are not very compatible with icecream.

<hr>

[1]Alas, the Python 2.7 support was one of the enticing features of icecream, but something ycecream lacks.

**Comment by jkyeung** [comment]
@gruns - I am sure I am in a very tiny minority, but I would be interested in a single-file `icecream.py`.

My reason for wanting a single file is related to the reason I don't just go ahead and use `ycecream.py` (since I actually am not particularly interested in the color features). I would love to be able to use icecream with an obscure and esoteric implementation of Python that is stuck at 2.7 and does not have pip or any other easy installers.

**Comment by alexmojaki** [comment]
> obscure and esoteric implementation of Python

What implementation? The magic behind icecream and ycecream uses a lot of CPython implementation details. PyPy imitates these details which is why it works there, but there's no guarantees for all Pythons.

Download https://github.com/alexmojaki/executing/blob/master/executing/executing.py (a single file!) and then try running this code (from a file, not a shell):

```python
import ast
import executing
import inspect

frame = inspect.currentframe()
print(ast.dump(executing.Source.for_frame(frame).executing(frame).node))
```

The expected output is something like this:

```python
Call(func=Attribute(value=Call(func=Attribute(value=Attribute(value=Name(id='executing', ctx=Load()), attr='Source', ctx=Load()), attr='for_frame', ctx=Load()), args=[Name(id='frame', ctx=Load())], keywords=[], starargs=None, kwargs=None), attr='executing', ctx=Load()), args=[Name(id='frame', ctx=Load())], keywords=[], starargs=None, kwargs=None)
```

If you just get an exception or `None`, icecream isn't going to work.

**Comment by alexmojaki** [comment]
Otherwise you can try https://github.com/zestyping/q . It comes in a single file and is slightly less magical so it might have a better chance of working outside of CPython.

**Comment by jkyeung** [comment]
_Originally posted by @salabim in https://github.com/gruns/icecream/issues/67#issuecomment-812521591_

> Ycvecream is now available under Python 2.7.
> No dependencies and the full functionality! Give it a try.

Excellent! I tried it and it works!

You got through the backport quicker than I expected.

**Comment by gruns** [comment]
closing

future reading of this issue: please open tickets/PRs for any features you'd like to see in icecream -- like from ycream or elsewhere!
